### DISCUSSION TITLE
Improve Image Uploading Experience

### DISCUSSION THREAD

**Discussion by steven-tey** [steven-tey/novel#36] [Ideas]
Today, we're launching Image Uploads for Novel:

https://github.com/steven-tey/novel/assets/28986134/b1406f79-8a09-42d2-9546-e6720f3e0200

Images are uploaded and stored in [Vercel Blob](https://vercel.com/storage/blob).

However, this is not the ideal UX..._yet_.

## Potential Improvement

Would love to know if it's possible to show the local base64 version of the image as a preview while the upload is in progress, and when it's done, replace the base64 url with the uploaded URL. 

This is similar to Notion's uploading experience:

https://github.com/steven-tey/novel/assets/28986134/7314871f-fe0e-415e-ab69-33e89a54d9e9

I scoured the web and can't seem to find a way to reliably change the `src` of an image node after the image is fully uploaded. The closes I could find is [this comment](https://gist.github.com/slava-vishnyakov/16076dff1a77ddaca93c4bccd4ec4521?permalink_comment_id=3458091#gistcomment-3458091) but I couldn't figure out how to make it work 😅 

Would love to know if other folks have any ideas on how to implement this!

**Comment by hernancito** [comment]
Hi Steven! I'm sharing a demo and the code used in our editor to preview the image before switching to the correct URL. It flashes in the swap as Guillermo tweeted https://twitter.com/rauchg/status/1671896959152701440?s=20 😅 but I think I know how to improve it by looking at the Novel repo
Hope this helps, I can do a PR later to give it a try.

https://github.com/steven-tey/novel/assets/18347863/1df1a73b-b20e-487b-a515-dd437d3be113

## A quick explanation
I did it by following this code https://gist.github.com/slava-vishnyakov/16076dff1a77ddaca93c4bccd4ec4521?permalink_comment_id=3768989 (below is the updated version to work with Tiptap 2.0 https://gist.github.com/slava-vishnyakov/16076dff1a77ddaca93c4bccd4ec4521#gistcomment-3744392)

1. We need a plugin to handle the Prosemirror events (cursor position, URL swapping) 

`UploadPastedImagesPlugin.tsx`

```javascript
import { Plugin, PluginKey } from '@tiptap/pm/state';
import { Decoration, DecorationSet } from '@tiptap/pm/view';

const uploadKey = new PluginKey('history');
export type UploadFn = (image: File) => Promise<any>;

const UploadPastedImagesPlugin = (upload: UploadFn) =>
  new Plugin({
    key: uploadKey,
    state: {
      init() {
        return DecorationSet.empty;
      },
      apply(tr, set) {
        // Adjust decoration positions to changes made by the transaction
        set = set.map(tr.mapping, tr.doc);
        // See if the transaction adds or removes any placeholders
        const action = tr.getMeta(this);
        if (action && action.add) {
          // Lower image opacity to indicate uploading process
          const widget = document.createElement('img');
          widget.setAttribute('class', 'opacity-40');
          widget.src = action.add.src;
          const deco = Decoration.widget(action.add.pos, widget, {
            id: action.add.id,
          });
          set = set.add(tr.doc, [deco]);
        } else if (action && action.remove) {
          set = set.remove(
            set.find(null, null, (spec) => spec.id == action.remove.id)
          );
        }
        return set;
      },
    },
    props: {
      decorations(state) {
        return this.getState(state);
      },
      handleDOMEvents: {
        paste(view, event) {
          pasteHandler(view, event, upload);
        },
      },
    },
  });

const pasteHandler = (view, event, upload) => {
  // Get the data of clipboard
  const clipboardItems = event?.clipboardData?.items;
  if (!clipboardItems) return false;
  const items = Array.from(clipboardItems).filter((item: any) => {
    // Filter the image items only
    return item.type.indexOf('image') !== -1;
  });
  if (items.length === 0) {
    return false;
  }

  const item: any = items[0];
  const file = item.getAsFile();
  if (!file) {
    return false;
  }
  if (event?.clipboardData?.types.includes('text/rtf')) {
    // Do not convert pasted rtf to image
    return false;
  }
  startImageUpload(view, file, upload);
  event.preventDefault();
  return true;
};

function findPlaceholder(state, id) {
  const decos = uploadKey.getState(state);
  const found = decos.find(null, null, (spec) => spec.id == id);
  return found.length ? found[0].from : null;
}

function startImageUpload(view, file, upload) {
  // A fresh object to act as the ID for this upload
  const id = {};

  // Replace the selection with a placeholder
  const tr = view.state.tr;
  if (!tr.selection.empty) tr.deleteSelection();

  const reader = new FileReader();
  reader.readAsDataURL(file);
  reader.onload = () => {
    tr.setMeta(uploadKey, {
      add: {
        id,
        pos: tr.selection.from,
        src: reader.result,
      },
    });
    view.dispatch(tr);
  };

  upload(file).then((src) => {
    const pos = findPlaceholder(view.state, id);
    // If the content around the placeholder has been deleted, drop
    // the image
    if (pos == null) return;
    // Otherwise, insert it at the placeholder's position, and remove
    // the placeholder
    view.dispatch(
      view.state.tr
        .replaceWith(pos, pos, view.state.schema.nodes.image.create({ src }))
        .setMeta(uploadKey, { remove: { id } })
    );
  });
}

export default UploadPastedImagesPlugin;
```

2. We will use this plugin to extend our Image component. It is almost a copy of [image.ts from tiptap](https://github.com/ueberdosis/tiptap/blob/develop/packages/extension-image/src/image.ts)

```javascript
import { Node, nodeInputRule } from '@tiptap/core';
import { mergeAttributes } from '@tiptap/react';
import UploadPastedImagesPlugin, { UploadFn } from './UploadPastedImagesPlugin';

interface ImageOptions {
  inline: boolean;
  HTMLAttributes: Record<string, any>;
}

const IMAGE_INPUT_REGEX = /!\[(.+|:?)\]\((\S+)(?:(?:\s+)["'](\S+)["'])?\)/;

const ImageExtension = (uploadFn: UploadFn) => {
  return Node.create<ImageOptions>({
    name: 'image',

    addOptions() {
      return {
        inline: false,
        HTMLAttributes: {},
      };
    },

    inline() {
      return this.options.inline;
    },

    group() {
      return this.options.inline ? 'inline' : 'block';
    },

    draggable: true,

    addAttributes() {
      return {
        src: {
          default: null,
        },
        alt: {
          default: null,
        },
        title: {
          default: null,
        },
      };
    },
    parseHTML: () => [
      {
        tag: 'img[src]',
        getAttrs: (dom) => {
          if (typeof dom === 'string') return {};
          const element = dom as HTMLImageElement;

          const obj = {
            src: element.getAttribute('src'),
            title: element.getAttribute('title'),
            alt: element.getAttribute('alt'),
          };
          return obj;
        },
      },
    ],
    renderHTML: ({ HTMLAttributes }) => [
      'img',
      mergeAttributes(HTMLAttributes),
    ],

    addCommands() {
      return {
        setImage:
          (attrs) =>
          ({ state, dispatch }) => {
            const { selection } = state;
            const position = selection.$head
              ? selection.$head.pos
              : selection.$to.pos;

            const node = this.type.create(attrs);
            const transaction = state.tr.insert(position, node);
            return dispatch?.(transaction);
          },
      };
    },
    addInputRules() {
      return [
        nodeInputRule({
          find: IMAGE_INPUT_REGEX,
          type: this.type,
          getAttributes: (match) => {
            const [, alt, src, title] = match;
            return {
              src,
              alt,
              title,
            };
          },
        }),
      ];
    },
    addProseMirrorPlugins() {
      return [UploadPastedImagesPlugin(uploadFn)];
    },
  });
};

export default ImageExtension;
```

3. Use the new ImageExtension in the editor 🎉:

```javascript
import ImageExtension from './ImageExtension';

const savePastedImage = async (file: File) => {
...
}

new Editor({
  extensions: [
    ...
    ImageExtension(savePastedImage),
    ...
```

**Reply by steven-tey** [reply]
Holy crap! That looks sublime! :amazing-rainbow: 

You could potentially load the uploaded image's URL before replacing the image in the DOM – similar to how we're doing it here: https://github.com/steven-tey/novel/blob/main/lib/editor.ts#L44-L50

Also, is this approach immune to the edge case where the user clicks on a different spot in the editor before the image finishes uploading? That tripped me up before when I tried to implement this 😅 

Would love to check out the PR!

**Reply by hernancito** [reply]
Yeah! The image retains the position where it was pasted, here a demo (I quickly jumped to the end of the text using the keyboard shortcut):

https://github.com/steven-tey/novel/assets/18347863/cc7ada1a-b54a-4286-84db-51e0328d32cf

I'll work on the PR

**Reply by steven-tey** [reply]
Incredible! 🤯

Can't wait to check out the PR! 🙏 

Also, that app you're building looks gorgeous! Amazing work 👏

**Reply by hernancito** [reply]
Thank you! I really appreciate your comment ❤️
I started working on the PR, adapting it to Novel

**Reply by steven-tey** [reply]
@hernancito can't wait! :star-struck_cat: 

Does the implementation support multiple images btw? (related issue: https://github.com/steven-tey/novel/issues/69)

**Reply by hernancito** [reply]
I think we have a decent implementation. Here are demo (throttling applied on purpose because I don't have a Vercel Blob token).

https://github.com/steven-tey/novel/assets/18347863/40091364-e073-4e48-98a8-b64679c37028

And here I pasted an image between two blocks of text.

https://github.com/steven-tey/novel/assets/18347863/523be4fa-75ee-4640-a86f-e7107b39e620

There is a lot of room for improvement, such as avoiding adding a new line when pasting the image into an empty block

**Reply by steven-tey** [reply]
Incredible work @hernancito! Just merged the PR, and announced it on [Twitter](https://twitter.com/steventey/status/1677706428629549057) / [LinkedIn](https://www.linkedin.com/posts/steventey_the-image-upload-flow-in-httpsnovelsh-activity-7083472165525716992-eGdR?utm_source=share&utm_medium=member_desktop)!

Thank YOU for the PR!

**Reply by hernancito** [reply]
I really appreciate your shoutout @steven-tey! Thank you for your great work too
